internal/template-validator/virtinformers/vmcache.go

```diff
@@ -0,0 +1,255 @@
+package virtinformers
+
+import (
+	"reflect"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	"kubevirt.io/ssp-operator/internal/template-validator/labels"
+)
+
+type VmCache interface {
+	cache.Store
+
+	HasSynced() bool
+
+	GetVmsForTemplate(template string) []string
+}
+
+type VmCacheValue struct {
+	Vm       string
+	Template string
+}
+
+func newVmCacheValue(obj metav1.Object) *VmCacheValue {
+	templateKeys := labels.GetTemplateKeys(obj)
+	return &VmCacheValue{
+		Vm:       vmCacheKey(obj),
+		Template: templateKeys.Get().String(),
+	}
+}
+
+func vmCacheKey(obj metav1.Object) string {
+	return types.NamespacedName{
+		Name:      obj.GetName(),
+		Namespace: obj.GetNamespace(),
+	}.String()
+}
+
+type NameSet = map[string]struct{}
+
+type templateMap map[string]NameSet
+
+func (t templateMap) Add(template, vm string) {
+	vmNameSet, exists := t[template]
+	if !exists {
+		vmNameSet = NameSet{}
+		t[template] = vmNameSet
+	}
+
+	vmNameSet[vm] = struct{}{}
+}
+
+func (t templateMap) Delete(template, vm string) {
+	vmNameSet, exists := t[template]
+	if !exists {
+		return
+	}
+
+	delete(vmNameSet, vm)
+	if len(vmNameSet) == 0 {
+		delete(t, template)
+	}
+}
+
+func (t templateMap) List(template string) []string {
+	names, exists := t[template]
+	if !exists {
+		return nil
+	}
+	result := make([]string, 0, len(names))
+	for name := range names {
+		result = append(result, name)
+	}
+	return result
+}
+
+type Predicate func(vm metav1.Object) bool
+
+type vmCache struct {
+	lock sync.RWMutex
+
+	store          map[string]VmCacheValue
+	vmsForTemplate templateMap
+	hasSynced      bool
+
+	filter Predicate
+}
+
+var _ VmCache = &vmCache{}
+
+func NewVmCache(filter Predicate) VmCache {
+	return &vmCache{
+		store:          map[string]VmCacheValue{},
+		vmsForTemplate: templateMap{},
+		filter:         filter,
+	}
+}
+
+func (v *vmCache) Add(obj interface{}) error {
+	metaObj, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	if !v.filter(metaObj) {
+		return nil
+	}
+
+	key := vmCacheKey(metaObj)
+	val := newVmCacheValue(metaObj)
+
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	v.store[key] = *val
+	v.vmsForTemplate.Add(val.Template, val.Vm)
+	return nil
+}
+
+func (v *vmCache) Update(obj interface{}) error {
+	newObjMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	if !v.filter(newObjMeta) {
+		return v.Delete(obj)
+	}
+
+	key := vmCacheKey(newObjMeta)
+	newVal := newVmCacheValue(newObjMeta)
+
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	oldVal, exists := v.store[key]
+	if !exists {
+		v.store[key] = *newVal
+		v.vmsForTemplate.Add(newVal.Template, newVal.Vm)
+		return nil
+	}
+
+	if reflect.DeepEqual(oldVal, newVal) {
+		return nil
+	}
+
+	v.store[key] = *newVal
+	v.vmsForTemplate.Delete(oldVal.Template, oldVal.Vm)
```

Comment:
Couldn't you add an update mechanism to `templateMap` instead of deleting and recreating?

Reply:
I considered it, but I couldn't think of a more efficient way how to implement update, than simply delete followed by add.
Can you think of something?

Reply:
Wouldn't `Add` already act as an update? Or does it also have to remove?
If so then yes, I think delete and recreate is probably best.

Reply:
The `templateMap` is a mapping from template name to a set of VM names. The update that I need is to change the template that is referenced by a VM, so update from `(template1, vm)` to `(template2, vm)`. But the template name is the key of the map, so the old key needs to be removed and the new key added.

I'm not sure I can explain it better that the code.

Reply:
Right, that's how I understood it. Then disregard my question. Good this way.